### ISSUE TITLE
delete WebAssembly

### ISSUE THREAD

**Issue by niutech** [stevespringett/disable-webassembly#6]
Couldn't you just invoke a JS command `delete WebAssembly` as a [userscript](https://en.wikipedia.org/wiki/Userscript)?

**Comment by minusmulticoloredper** [comment]
It’s quite difficult to achieve this. I tried using AI to create a user script aimed at blocking WebAssembly objects and .wasm file requests. After adding the script and visiting the WASM detection site (https://wasm.joway.io/), the website shows a message saying the browser doesn’t support WASM. However, when I refresh the page, WASM successfully loads before the user script is injected, and the website then indicates that the browser supports WASM.

Interestingly, when I disable caching through the browser’s developer console, no matter how many times I refresh the page, the site continuously shows the "browser doesn’t support WASM" message. I’m not sure if AI didn’t take this situation into account.

```javascript
// ==UserScript==
// @name         WebAssembly Blocker
// @namespace    http://tampermonkey.net/
// @version      1.0
// @description  Block WebAssembly features and .wasm file loading
// @author       DeepSeek R1
// @match        *://*/*
// @run-at       document-start
// @grant        none
// ==/UserScript==

(function() {
    'use strict';

    // Block WebAssembly object
    Object.defineProperty(window, 'WebAssembly', {
        configurable: false,
        writable: false,
        value: null
    });

    // Intercept .wasm file requests via XHR
    const originalOpen = XMLHttpRequest.prototype.open;
    XMLHttpRequest.prototype.open = function(method, url) {
        if (url && url.endsWith('.wasm')) {
            console.log('[Blocked] WASM file request:', url);
            return;
        }
        originalOpen.apply(this, arguments);
    };

    // Intercept fetch requests for WASM
    const originalFetch = window.fetch;
    window.fetch = function(input, init) {
        if (typeof input === 'string' && input.endsWith('.wasm')) {
            console.log('[Blocked] WASM fetch:', input);
            return Promise.reject(new Error('WASM blocked'));
        }
        return originalFetch.call(this, input, init);
    };

    // Prevent inline WASM instantiation
    if (window.WebAssembly && window.WebAssembly.instantiate) {
        window.WebAssembly.instantiate = function() {
            throw new Error('WebAssembly is disabled by user script');
        };
    }
})();
```

**Comment by daslicht** [comment]
> It’s quite difficult to achieve this. I tried using AI to create a user script aimed at blocking WebAssembly objects and .wasm file requests. After adding the script and visiting the WASM detection site (https://wasm.joway.io/), the website shows a message saying the browser doesn’t support WASM. However, when I refresh the page, WASM successfully loads before the user script is injected, and the website then indicates that the browser supports WASM.
> 
> Interestingly, when I disable caching through the browser’s developer console, no matter how many times I refresh the page, the site continuously shows the "browser doesn’t support WASM" message. I’m not sure if AI didn’t take this situation into account.
> 
> // ==UserScript==
> // @name         WebAssembly Blocker
> // @namespace    http://tampermonkey.net/
> // @version      1.0
> // @description  Block WebAssembly features and .wasm file loading
> // @author       DeepSeek R1
> // @match        *://*/*
> // @run-at       document-start
> // @grant        none
> // ==/UserScript==
> 
> (function() {
>     'use strict';
> 
>     // Block WebAssembly object
>     Object.defineProperty(window, 'WebAssembly', {
>         configurable: false,
>         writable: false,
>         value: null
>     });
> 
>     // Intercept .wasm file requests via XHR
>     const originalOpen = XMLHttpRequest.prototype.open;
>     XMLHttpRequest.prototype.open = function(method, url) {
>         if (url && url.endsWith('.wasm')) {
>             console.log('[Blocked] WASM file request:', url);
>             return;
>         }
>         originalOpen.apply(this, arguments);
>     };
> 
>     // Intercept fetch requests for WASM
>     const originalFetch = window.fetch;
>     window.fetch = function(input, init) {
>         if (typeof input === 'string' && input.endsWith('.wasm')) {
>             console.log('[Blocked] WASM fetch:', input);
>             return Promise.reject(new Error('WASM blocked'));
>         }
>         return originalFetch.call(this, input, init);
>     };
> 
>     // Prevent inline WASM instantiation
>     if (window.WebAssembly && window.WebAssembly.instantiate) {
>         window.WebAssembly.instantiate = function() {
>             throw new Error('WebAssembly is disabled by user script');
>         };
>     }
> })();

That doesn't work for me in safari, it still says supported

**Comment by Anonymous-Humanoid** [comment]
> Couldn't you just invoke a JS command `delete WebAssembly` as a [userscript](https://en.wikipedia.org/wiki/Userscript)?

Assuming this were doable without worry of race conditions or extension permissions, this still isn't a substitute for disabling WASM. Every website is executed in its own "world," basically a sandbox that is given access to the DOM and certain JavaScript APIs, depending on whether or not it's in a [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts). However, extension content scripts can inject code in an [isolated world](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#isolated_world), which means it's also given access to the DOM and some JS APIs.

The key takeaway here is that a website, userscript or extension can modify global variables all they want in one world, but it won't affect any isolated worlds, and vice versa. Meaning WebAssembly can still potentially be accessed, albeit from trusted sources. Unless they suffer from a [compromised renderer](https://chromium.googlesource.com/chromium/src/+/master/docs/security/compromised-renderers.md), or other exposure of potentially privileged APIs.

**Comment by niutech** [comment]
@Anonymous-Humanoid How is it so? When one script deletes global `window.WebAssembly`, how would it be available in another script in the same HTML document?

**Comment by Anonymous-Humanoid** [comment]
@niutech I'm happy to share! TL;DR it has nothing to do with the HTML document, but rather that extension content scripts can run in isolated environments that have access to `window.WebAssembly`, even if it's deleted in the main one the website shares.

## Overview

From the [Chrome extension docs](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts#isolated_world):

> Content scripts live in an isolated world, allowing a content script to make changes to its JavaScript environment without conflicting with the page or other extensions' content scripts.

Essentially, every web page has multiple "worlds", which are isolated JavaScript environments. Extension content scripts can run in the main world, sharing the JavaScript of the page, or in an isolated world. Isolated worlds can still access the DOM just the same as the main world, but they cannot be affected by scripts in other worlds.

## Visual representation

If you'll excuse the crude penmanship, here's a simplified visual explanation of a single tab process:

![Diagram showing how content scripts, worlds, the DOM and browser APIs interact](https://github.com/user-attachments/assets/0e3de695-b458-43d6-b0a5-dd5053f444ee)

Content scripts default to being injected into the main world, where it shares the JavaScript of the main page, but it won't affect any other isolated worlds, because they're all initially given access to the same JavaScript APIs. Basically, each world creates global variables, such as `window.WebAssembly`, that point to immutable APIs. The global variables themselves are mutable, but reassigning them doesn't affect the internal APIs. This is how scripts can achieve isolation from one another, which avoids conflicting variables or interference. In simpler terms, every world initially gets its own copy of the global variables. For completeness, this diagram also shows that the DOM is shared throughout worlds, but it also has access to functions defined in the main world.

## Practical demo

Below is a sample Chrome extension to practically demonstrate these concepts.

`package.json`:

```json
{
    "devDependencies": {
        "@types/chrome": "^0.0.323"
    }
}
```

`manifest.json`:

```json
{
    "name": "sample-extension",
    "version": "1.0.0",
    "manifest_version": 3,
    "permissions": ["activeTab", "scripting"],
    "action": {},
    "background": {
        "service_worker": "background.js"
    }
}
```

`background.js`:

```ts
/** @typedef {chrome.scripting.InjectionResult<void>[]} InjectionResult */

/** @type {((tabId: number) => Promise<InjectionResult>)[]} */
const contentScripts = [];

/**
 * @param {chrome.scripting.ExecutionWorld} world
 * @param {() => void} func 
 */
function addContentScript(world, func) {
    contentScripts.push((tabId) => {
        return chrome.scripting.executeScript({
            target: {
                tabId,
                allFrames: true
            },
            world,
            func
        });
    });
}

/**
 * @param {number} ms The time to sleep in milliseconds
 * @returns {Promise<void>} A promise that resolves after the specified time
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Initializing content scripts sequentially
addContentScript('MAIN', () => {
    console.log('Running first script...');
    console.log(window.WebAssembly);
    // @ts-expect-error
    delete window.WebAssembly;
    console.log(window.WebAssembly);
    console.log('Finished first script!');
});

addContentScript('MAIN', () => {
    console.log('Running second script...');
    console.log(window.WebAssembly);
    // @ts-expect-error
    delete window.WebAssembly;
    console.log(window.WebAssembly);
    console.log('Finished second script!');
});

addContentScript('ISOLATED', () => {
    console.log('Running third script...');
    console.log(window.WebAssembly);
    console.log('Finished third script!');
});

// Running content scripts sequentially
chrome.action.onClicked.addListener(async (tab) => {
    if (tab.id != null) {
        for (let script of contentScripts) {
            await script(tab.id);
            await sleep(1000); // Waiting to show there are no race conditions
        }
    }
});
```

When you navigate to an HTTP(S) page and click on the extension icon, the developer console for that tab will show the program output. In the below screenshot, I did this with [example.com](https://example.com/):

![First script removes WASM; second script and main page can't access WASM; third script can access WASM](https://github.com/user-attachments/assets/3d7adc3a-2419-46e4-8dab-7ee99c91c8a7)

Let's break this down. For the sake of explanation, the only important part is this part of the background script:

```ts
// Initializing content scripts sequentially
addContentScript('MAIN', () => {
    console.log('Running first script...');
    console.log(window.WebAssembly);
    // @ts-expect-error
    delete window.WebAssembly;
    console.log(window.WebAssembly);
    console.log('Finished first script!');
});

addContentScript('MAIN', () => {
    console.log('Running second script...');
    console.log(window.WebAssembly);
    // @ts-expect-error
    delete window.WebAssembly;
    console.log(window.WebAssembly);
    console.log('Finished second script!');
});

addContentScript('ISOLATED', () => {
    console.log('Running third script...');
    console.log(window.WebAssembly);
    console.log('Finished third script!');
});
```

When running the extension, the first content script runs in the main world, sharing the same JavaScript as the page. It outputs the WASM global variable, deletes it and outputs it again to show that it was, in fact, deleted.

> Running first script...
>
> *WebAssembly {...}*
>
> *undefined*
>
> Finished first script!

Once completed, the second content script does the same thing in the main world, showing that other scripts in the main world are also affected by this change. After all the scripts ran, I also manually asserted through the developer console that `window.WebAssembly` isn't defined to further illustrate this.

> Running second script...
>
> *undefined*
>
> *undefined*
>
> Finished second script!

Finally, the third content script runs in its own isolated world and just outputs the global variable without affecting it.

> Running third script...
>
> *WebAssembly {...}*
>
> Finished third script!

What this shows is that the main world cannot affect any isolated world, which means they can't delete the global variable everywhere it can possibly be used.

## Conclusion

Deleting any global variable is not sufficient to prevent it being exposed to other content scripts, as isolated worlds are provided with new global variables to access internal JavaScript APIs without external script interference. For security reasons, preventing these APIs from being exposed requires modifying browser settings, rather than installing an extension.

However, if the goal of disabling WASM is to increase privacy rather than just security, then deleting `window.WebAssembly` would be sufficient if you could guarantee that occurs before any other scripts run in the main world- which you cannot. I say this, because you should trust any extensions installed, as they have higher privileges than any website can access, even with the user's full cooperation.

EDIT: Realizing now that I sound like AI. Not happy about that.
